src/train_tracker.py

```diff
@@ -0,0 +1,228 @@
+            cv2.imshow("Calibration", mark_frame)
+            # フレームのファイル出力
+            video.write(frame)              # 撮影した動画
+            mark_video.write(mark_frame)    # 輪郭線を描画した動画
+
+            # "q"キーを押すとループを抜けて終了
+            if cv2.waitKey(1) & 0xFF == ord("q"):
+                break
+
+        # リソースを解放
+        self.camera.end_record()
+
+    def mouse_callback(self, event, x, y, flags, param) -> None:
+        """キャリブレーションのクリック時の処理.
+
+        Args:
+            event: クリックイベント
+            x: クリックのx座標
+            y: クリックのy座標
+            flags: 追加情報のフラグ
+            param: 任意パラメータ
+        """
+        if event == cv2.EVENT_LBUTTONDOWN:
+            # クリック時の座標を記録
+            self.observe_rect_points.append((x, y))
+
+    def draw_observe_rect(self, frame) -> np.ndarray:
+        """監視する領域をフレームに描画する.
+
+        Args:
+            frame: 描画対象フレーム
+        Returns:
+            mark_frame: 監視する領域を描画したフレーム
+        """
+        mark_frame = frame.copy()
+        cv2.rectangle(mark_frame, self.observe_rect_points[-1],
+                      self.observe_rect_points[-2], (0, 255, 255), 2)
+        return mark_frame
+
+    def observe(self) -> None:
+        """映像を監視する."""
+        # 録画を開始する
+        video, mark_video = self.camera.start_record()
+
+        initial_frame = None
+        # フレームを取得して動画に書き込む
+        while True:
+            # フレームを取得
+            success, frame = self.camera.get_frame()
+            if not success:
+                break
+
+            # 初期フレームを保持
+            if initial_frame is None:
+                gray_frame = cv2.cvtColor(frame, cv2.COLOR_BGR2GRAY)
+                initial_frame = gray_frame.copy().astype("float")
+                # 以降の処理をスキップし次のループへ
+                continue
+            # 動体の検出
+            mark_frame = self.detect_motion(frame, initial_frame)
+            # 列車の検出
+            mark_frame, train_rect_points = self.detect_train(mark_frame)
+            # クリックされた領域の描画
+            mark_frame = self.draw_observe_rect(mark_frame)
+
+            if len(train_rect_points) >= 2:
+                # 列車の境界
+                (train_left, train_top), (train_right,
+                                          train_bottom) = train_rect_points
+                # 描画した指定領域
+                observe_y_values = [
+                    y for x, y in self.observe_rect_points[-2:]]
+                observe_x_values = [
+                    x for x, y in self.observe_rect_points[-2:]]
+
+                observe_top = min(observe_y_values)
+                observe_bottom = max(observe_y_values)
+                observe_left = min(observe_x_values)
+                observe_right = max(observe_x_values)
+
+                # 列車が指定領域に侵入しているかを判定する.
+                # 列車の先頭(下方向)の動画内の高さが指定領域内
+                if observe_top < train_bottom < observe_bottom:
+                    # 列車の動画内の横幅が指定領域内
+                    if observe_left < train_left \
+                            and train_right < observe_right:
```

Comment:
```suggestion
                    # 列車がobserve_leftより右側
                    if observe_left < train_left \
                            and observe_left < train_right:
```
observe_rightより右にあって困ることはない気がする。

Reply:
Lコース